src/dbnode/storage/shard.go

```diff
@@ -2849,9 +2816,99 @@ READER:
 
 	s.logger.Debug("finished aggregating tiles",
 		zap.Uint32("shard", s.ID()),
-		zap.Int64("processedTiles", processedTileCount.Load()))
+		zap.Int64("processedTiles", processedTileCount))
+
+	return processedTileCount, nil
+}
+
+func encodeAggregatedSeries(
+	seriesIter tile.SeriesFrameIterator,
+	annotationPayload annotation.Payload,
+	downsampledValues []downsampled.Value,
+	encoder encoding.Encoder,
+) (int64, error) {
+	var (
+		prevFrameLastValue = math.NaN()
+		processedTileCount int64
+		handleValueResets  bool
+		firstUnit          xtime.Unit
+		firstAnnotation    ts.Annotation
+		err                error
+	)
+
+	for seriesIter.Next() {
+		frame := seriesIter.Current()
 
-	return processedTileCount.Load(), nil
+		frameValues := frame.Values()
+		if len(frameValues) == 0 {
+			continue
+		}
+
+		if processedTileCount == 0 {
+			firstUnit, err = frame.Units().Value(0)
+			if err != nil {
+				return 0, err
+			}
+
+			firstAnnotation, err = frame.Annotations().Value(0)
```

Comment:
nit: may be better to add a `First()` method to these interfaces? Don't think it even needs an error since we should be fine after `if len(frameValues) == 0` check above

Reply:
This idea had crossed my mind, but I didn't like it because of too much special-casing, also that would make the universal `Value(idx)` API dead code effectively.